src/internal/storage/fileset/writer.go

```diff
@@ -102,80 +101,82 @@ func (w *Writer) checkPath(prevIdx, idx *index.Index) error {
 	return nil
 }
 
-// Copy copies a file to the file set writer.
-func (w *Writer) Copy(file File, datum string) error {
-	idx := file.Index()
-	copyIdx := &index.Index{
-		Path: idx.Path,
+// Delete creates a delete operation for a file.
+func (w *Writer) Delete(path, datum string) error {
+	idx := &index.Index{
+		Path: path,
 		File: &index.File{
 			Datum: datum,
 		},
 	}
-	if err := w.nextIdx(copyIdx); err != nil {
+	if err := w.checkIndex(w.deleteIdx, idx); err != nil {
 		return err
 	}
-	// Copy the file data refs.
-	for _, dataRef := range idx.File.DataRefs {
-		w.sizeBytes += dataRef.SizeBytes
-		if err := w.cw.Copy(dataRef); err != nil {
-			return err
-		}
-	}
-	return nil
+	w.deleteIdx = idx
+	return w.deletive.WriteIndex(idx)
 }
 
-func (w *Writer) callback(annotations []*chunk.Annotation) error {
-	for _, annotation := range annotations {
-		idx := annotation.Data.(*index.Index)
-		if w.lastIdx == nil {
-			w.lastIdx = idx
-		}
-		if idx.Path != w.lastIdx.Path || idx.File.Datum != w.lastIdx.File.Datum {
-			if err := w.additive.WriteIndex(w.lastIdx); err != nil {
-				return err
-			}
-			if w.indexFunc != nil {
-				if err := w.indexFunc(w.lastIdx); err != nil {
-					return err
-				}
-			}
-			w.lastIdx = idx
+// Copy copies a file to the file set writer.
+func (w *Writer) Copy(file File, datum string) error {
+	idx := file.Index()
+	size := index.SizeBytes(idx)
+	if size >= int64(w.batchThreshold) && chunk.StableDataRefs(idx.File.DataRefs) {
+		if err := w.checkIndex(w.idx, idx); err != nil {
+			return err
 		}
-		if annotation.NextDataRef != nil {
-			w.lastIdx.File.DataRefs = append(w.lastIdx.File.DataRefs, annotation.NextDataRef)
+		w.idx = idx
+		copyIdx := &index.Index{
+			Path: idx.Path,
+			File: &index.File{
+				Datum: datum,
+			},
 		}
-	}
-	return nil
+		return w.uploader.Copy(copyIdx, idx.File.DataRefs)
+	}
+	// TODO: Optimize to handle large files that can mostly be copy by reference?
```

Comment:
for the below-threshold case or supporting older large files? I don't see how you can do a copy by reference in those cases with the current model

Reply:
Large files are spread across multiple file sets during upload, so a compaction would rewrite them. I am thinking that is fine for now because it is only once, but it is less than ideal.